src/main/java/com/aws/greengrass/clientdevices/auth/PermissionEvaluationUtils.java

```diff
@@ -56,118 +55,170 @@ public PermissionEvaluationUtils(GroupManager groupManager) {
             return true;
         }
-        if (String.format(SERVICE_OPERATION_FORMAT, requestOperation.getService(), ANY_REGEX).equals(policyOperation)) {
+        if (Objects.equals(requestOperation.getService() + DELIM + WILDCARD, policyOperation)) {
             return true;
         }
-        return ANY_REGEX.equals(policyOperation);
+        return Objects.equals(WILDCARD, policyOperation);
     }
 
     private boolean compareResource(Resource requestResource, String policyResource) {
-        if (requestResource.getResourceStr().equals(policyResource)) {
+        if (Objects.equals(requestResource.getResourceStr(), policyResource)) {
             return true;
         }
-
-        if (String.format(SERVICE_RESOURCE_FORMAT, requestResource.getService(), requestResource.getResourceType(),
-                ANY_REGEX).equals(policyResource)) {
+        if (Objects.equals(
+                requestResource.getService() + DELIM + requestResource.getResourceType() + DELIM + WILDCARD,
+                policyResource)) {
             return true;
         }
-
-        return ANY_REGEX.equals(policyResource);
+        return Objects.equals(WILDCARD, policyResource);
     }
 
-    private Operation parseOperation(String operationStr) {
+    private Operation parseOperation(String operationStr) throws PolicyException {
         if (Utils.isEmpty(operationStr)) {
-            throw new IllegalArgumentException("Operation can't be empty");
+            throw new PolicyException("Operation can't be empty");
+        }
+
+        int split = operationStr.indexOf(':');
+        if (split == -1 || split == operationStr.length() - 1) {
+            throw new PolicyException(EXCEPTION_MALFORMED_OPERATION);
+        }
+
+        String service = operationStr.substring(0, split);
+        if (service.isEmpty() || !StringUtils.isAlpha(service)) {
+            throw new PolicyException(EXCEPTION_MALFORMED_OPERATION);
         }
 
-        Matcher matcher = SERVICE_OPERATION_PATTERN.matcher(operationStr);
-        if (matcher.matches()) {
-            return Operation.builder().operationStr(operationStr).service(matcher.group(1)).action(matcher.group(2))
-                    .build();
+        String action = operationStr.substring(split + 1);
+        if (action.isEmpty() || !isAlphanumericWithExtraChars(action, "-_")) {
+            throw new PolicyException(EXCEPTION_MALFORMED_OPERATION);
         }
-        throw new IllegalArgumentException(String.format("Operation %s is not in the form of %s", operationStr,
-                SERVICE_OPERATION_PATTERN.pattern()));
+
+        return Operation.builder()
+                .operationStr(operationStr)
+                .service(service)
+                .action(action)
+                .build();
     }
 
-    private Resource parseResource(String resourceStr) {
+    private Resource parseResource(String resourceStr) throws PolicyException  {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 178-208 into a separate method.